hazelcast/src/main/java/com/hazelcast/client/impl/spi/ProxyManager.java

```diff
@@ -330,13 +388,64 @@ private void initialize(ClientProxy clientProxy) throws Exception {
     }
 
     public Collection<? extends DistributedObject> getDistributedObjects() {
+        ClientMessage request = ClientGetDistributedObjectsCodec.encodeRequest();
+        distributedObjectsWriteLock.lock();
+        try {
+            ClientInvocationFuture future = new ClientInvocation(client, request, client.getName()).invoke();
+            ClientMessage response = future.get();
+            processDistributedObjectInfos(response);
+        } catch (Exception e) {
+            throw rethrow(e);
+        } finally {
+            distributedObjectsWriteLock.unlock();
+        }
+
+        return getLocalDistributedObjects();
+    }
+
+    public Collection<? extends DistributedObject> getLocalDistributedObjects() {
         Collection<DistributedObject> objects = new LinkedList<>();
         for (ClientProxyFuture future : proxies.values()) {
             objects.add(future.get());
         }
         return objects;
     }
 
+    class SyncDistributedObjectsTask implements Runnable {
+        @Override
+        public void run() {
+            ClientMessage request = ClientGetDistributedObjectsCodec.encodeRequest();
+            distributedObjectsWriteLock.lock();
+            ClientInvocationFuture future = new ClientInvocation(client, request, client.getName()).invoke();
+            future.whenCompleteAsync((clientMessage, throwable) -> {
+                try {
+                    processDistributedObjectInfos(clientMessage);
+                } finally {
+                    distributedObjectsWriteLock.unlock();
```

Comment:
This will not work as you are doing the unlock in a different thread. 
You should wait for future's response on this thread and unlock on this thread.